changefeedccl: unify initial_scan option syntax

sherman-grewal opened the pull request:
Resolves #79324

Currently, we have explicit options for each possible
behaviour that a user would like to achieve for
initial scans on changefeeds. For instance, a user
could specify:

- initial_scan
- no_initial_scan
- initial_scan_only

This seems a bit sprawling, and can inadvertently cause
contradictions in a changefeed statement. Hence, in this
PR we extend the option `initial_scan` to take on three
possible values: `'yes|no|only'`. Once this change
is made we will remove the explicit options from the
docs, but we will keep these options for backwards
compatibility.

Release note (enterprise change): Unify the syntax that
allows users to define the behaviour they would like
for initial scans on changefeeds by extending the
`initial_scan` option to take on three possible values:
`'yes|no|only'`.

Release justification: Small, safe refactor that will
improve the user experience when creating changefeeds.

Jira issue: CRDB-14693

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79324)
<!-- Reviewable:end -->

miretskiy commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -97,6 +113,10 @@ const (
 
+	OptNoInitialScanType   InitialScanOptType = `no`
+	OptInitialScanType     InitialScanOptType = `yes`
+	OptInitialScanOnlyType InitialScanOptType = `only`
```
do we really need these?

miretskiy replied:
Yeah; just parse ... strings.ToLower()....

miretskiy replied:
I guess... You could keep it to be consistent... But I just don't know how much value this provides.  We parse strings into already descriptive enum value -- and so, having a public option representing string value... probably less useful.

sherman-grewal replied:
yeah I can see how these variables can be a bit redundant, I'll remove them

miretskiy commented on pkg/ccl/changefeedccl/changefeed.go:
```diff
@@ -97,11 +97,32 @@ func makeSpansToProtect(
-	return (cursor && initialScan) || (!cursor && !noInitialScan)
+	initialScanType, initialScanSet := opts[changefeedbase.OptInitialScan]
+	_, initialScanOnlySet := opts[changefeedbase.OptInitialScanOnly]
+	_, noInitialScanSet := opts[changefeedbase.OptNoInitialScan]
```
presumably, this relies on validation done prior to ensure that both initial+no_initial are not set?

sherman-grewal replied:
yeah it does, I will move the validation checks into this function.

miretskiy commented on pkg/ccl/changefeedccl/changefeed.go:
```diff
@@ -97,11 +97,32 @@ func makeSpansToProtect(
-	_, initialScan := opts[changefeedbase.OptInitialScan]
-	_, noInitialScan := opts[changefeedbase.OptNoInitialScan]
-	return (cursor && initialScan) || (!cursor && !noInitialScan)
+	initialScanType, initialScanSet := opts[changefeedbase.OptInitialScan]
```
I wonder if we should do something like (to make sure parsing is case insensitive).
```
switch strings.ToLower(opts[changefeedbase.OptInitialScan])) {
   case 'yes': return ...
}

// OptInitialScan wasn't set using yes/no/only
.if no_initial_scan ....

````

miretskiy commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -763,13 +767,43 @@ func validateDetails(details jobspb.ChangefeedDetails) (jobspb.ChangefeedDetails
+			default:
+				return jobspb.ChangefeedDetails{}, errors.Errorf(
+					`unknown %s: %s`, opt, v)
+			}
```
I wonder if we should just change `initialScanFromOptions` methods to return InitialScan and an error.  Do some of the checking there... Maybe all of it, actually.

sherman-grewal replied:
yeah I think that might be a better option. Originally I didn't give it much thought because we always perform `validateDetails` before looking at this value, but these validations best belong in the function call.

Commit:
changefeedccl: unify initial_scan option syntax
Currently, we have explicit options for each possible
behaviour that a user would like to achieve for
initial scans on changefeeds. For instance, a user
could specify:

- initial_scan
- no_initial_scan
- initial_scan_only

This seems a bit sprawling, and can inadvertently cause
contradictions in a changefeed statement. Hence, in this
PR we extend the option `initial_scan` to take on three
possible values: `'yes|no|only'`. Once this change
is made we will remove the explicit options from the
docs, but we will keep these options for backwards
compatibility.

Release note (enterprise change): Unify the syntax that
allows users to define the behaviour they would like
for initial scans on changefeeds by extending the
`initial_scan` option to take on three possible values:
`'yes|no|only'`.

Release justification: Small, safe refactor that will
improve the user experience when creating changefeeds.

sherman-grewal commented:
bors r+

sherman-grewal commented:
bors r-

sherman-grewal commented:
bors r+
